lib/charms/mongodb/v0/mongos.py

```diff
@@ -235,11 +216,37 @@ def _is_shard_draining(self, shard_name: str) -> bool:
             f"Shard {shard_name} not in cluster, could not retrieve draining status"
         )
 
-    def get_primary_shard(self) -> str:
-        """Processes sc_status and identifies the primary shard."""
-        # TODO Follow up PR, implement this function there is no MongoDB command to retrieve
-        # primary shard, this is possible with mongosh.
-        return False
+    def get_primary_shards(self) -> Optional[List[str]]:
+        """Returns the primary shards of the cluster.
+
+        In Sharded MongoDB clusters, mongos selects the primary shard when creating a new database
+        by picking the shard in the cluster that has the least amount of data. This means that:
+        1. There can be multiple primary shards in a cluster.
+        2. Until there is data written to the cluster there is effectively no primary shard.
+        """
+        databases_collection = self._get_databases_collection()
+        if databases_collection is None:
+            logger.info("No data written to sharded cluster yet, no primary shards.")
+            return None
+
+        return databases_collection.distinct("primary")
+
+    def get_databases_for_primary_shard(self, primary_shard) -> Optional[List[str]]:
+        """Returns a list of databases using the given shard as a primary shard."""
+        databases_collection = self._get_databases_collection()
+        if databases_collection is None:
+            return
+
+        return databases_collection.distinct("_id", {"primary": primary_shard})
+
+    def _get_databases_collection(self) -> collection.Collection:
+        """Returns the database collection if present."""
+        config_db = self.client["config"]
+        if "databases" not in config_db.list_collection_names():
+            logger.info("No data written to sharded cluster yet.")
+            return None
+
+        return config_db["databases"]
```

Comment:
here we are returning collection by accessing `config_db["databases"]`  just curious why do we need a call to `config_db.list_collection_names()` before what

Reply:
`config_db.list_collection_names()` - lists all the collections present in the `config` database
`config_db["databases"]` - retrieves the collection named `databases` (which is not guaranteed to be present, this collection only gets created once data gets written to the cluster. I can add this to the docstring.)

Reply:
i am just curious is it possible to do something like

if "databases" in config_db without calling to `config_db.list_collection_names()` (since we are returning it like config_db['databases'].

If not - no problem 😄

Reply:
Sadly no, since `config_db ` is not an iterator we cannot do `config_db['databases']`, I tried something like this at first and got:

```
  File "/var/lib/juju/agents/unit-config-server-one-0/charm/lib/charms/mongodb/v0/mongos.py", line 245, in get_primary_shards
    if "databases" not in config_db:
TypeError: argument of type 'Database' is not iterable
```

Reply:
Thank you for the explanations :)

Reply:
of course! Thank you for your careful attention to detail!